File: benchmarks/cases/ackermann/ackermann.dart
// Ackermann–Péter function — deep, non-tail mutual self-recursion.

int add(int a, int b) => a + b;
int sub(int a, int b) => a - b;

int ack(int m, int n) {
  if (m == 0) return add(n, 1);
  if (n == 0) return ack(sub(m, 1), 1);
  return ack(sub(m, 1), ack(m, sub(n, 1)));
}

void main() {
  print(ack(3, 10));
}

File: benchmarks/cases/ackermann/ackermann.cs
// Ackermann–Péter function — deep, non-tail mutual self-recursion.

System.Console.WriteLine(Ack(3, 10));

static long Add(long a, long b) => a + b;
static long Sub(long a, long b) => a - b;

static long Ack(long m, long n)
{
    if (m == 0) return Add(n, 1);
    if (n == 0) return Ack(Sub(m, 1), 1);
    return Ack(Sub(m, 1), Ack(m, Sub(n, 1)));
}
